### PR TITLE
Align RPC with geth

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why this change is needed

- align some RPC methods with the expected geth results
- in preparation for a more rigorous OG RPC 

### What changes were made as part of this PR

- remove the custom "nil response" method and align with geth. Also adjust the "Await transaction" logic  (which should be consolidated)
- use the block or hash type
- change the receipt response
- change the transaction response
- adjust the input parameters 

### PR checks pre-merging

Please indicate below by ticking the checkbox that you have read and performed the required
[PR checks](https://github.com/ten-protocol/ten-internal/blob/main/dev-ops-docs/dev-pr-checks.md)

- [x] PR checks reviewed and performed

### CODE REVIEW

File: go/common/gethencoding/geth_encoding.go [hunk 1]
```diff
@@ -145,33 +144,68 @@ func ExtractAddress(param interface{}) (*gethcommon.Address, error) {
 }
 
 // ExtractOptionalBlockNumber defaults nil or empty block number params to latest block number
-func ExtractOptionalBlockNumber(params []interface{}, idx int) (*gethrpc.BlockNumber, error) {
+func ExtractOptionalBlockNumber(params []interface{}, idx int) (*gethrpc.BlockNumberOrHash, error) {
```
Comment (tudor-malene) [inline go/common/gethencoding/geth_encoding.go:147]: the geth api can send either a block number or a hash. 
We currently only support the former, but this change creates the plumbing for both

File: go/enclave/rpc/GetTransactionReceipt.go [hunk 2]
```diff
@@ -13,7 +17,7 @@ import (
 	"github.com/ten-protocol/go-ten/go/enclave/events"
 )
 
-func GetTransactionReceiptValidate(reqParams []any, builder *CallBuilder[gethcommon.Hash, types.Receipt], _ *EncryptionManager) error {
+func GetTransactionReceiptValidate(reqParams []any, builder *CallBuilder[gethcommon.Hash, map[string]interface{}], _ *EncryptionManager) error {
```
Comment (tudor-malene) [inline go/enclave/rpc/GetTransactionReceipt.go:20]: the geth api actually returns a map when you ask for receipts.

File: go/obsclient/authclient.go [hunk 3]
```diff
@@ -2,6 +2,8 @@ package obsclient
```
Comment (tudor-malene) [inline go/obsclient/authclient.go:2]: changes in this file are aligning with geth

File: go/rpc/encrypted_client.go [hunk 4]
```diff
@@ -235,7 +225,7 @@ func (c *EncRPCClient) executeSensitiveCall(ctx context.Context, result interfac
 
 	// If there is no encrypted response then this is equivalent to nil response
 	if rawResult.EncUserResponse == nil || len(rawResult.EncUserResponse) == 0 {
-		return ErrNilResponse
+		return nil
```
Comment (tudor-malene) [inline go/rpc/encrypted_client.go:228]: that error has no equivalent in geth

File: go/enclave/rpc/EstimateGas.go [hunk 5]
```diff
@@ -47,7 +47,8 @@ func EstimateGasValidate(reqParams []any, builder *CallBuilder[CallParamsWithBlo
 	}
 
 	builder.From = callMsg.From
-	builder.Param = &CallParamsWithBlock{callMsg, blockNumber}
+	// todo
```
Comment (zkokelj) [inline go/enclave/rpc/EstimateGas.go:50]: nitpick: What needs to be done here or can be this todo removed?

File: go/enclave/rpc/GetTransaction.go [hunk 6]
```diff
@@ -56,7 +56,7 @@ func GetTransactionExecute(builder *CallBuilder[gethcommon.Hash, RpcTransaction]
 	// Unlike in the Geth impl, we hardcode the use of a London signer.
 	// todo (#1553) - once the enclave's genesis.json is set, retrieve the signer type using `types.MakeSigner`
 	signer := types.NewLondonSigner(tx.ChainId())
-	rpcTx := newRPCTransaction(tx, blockHash, blockNumber, index, gethcommon.Big0, signer)
+	rpcTx := newRPCTransaction(tx, blockHash, blockNumber, index, rpc.config.BaseFee, signer)
```
Comment (zkokelj) [inline go/enclave/rpc/GetTransaction.go:59]: Out of curiosity.. do we want to have it configurable? Before it was hardcoded to 0?

**Review by zkokelj — APPROVED** [review]
LGTM 👍 , 2 minor comments added

File: go/enclave/rpc/EstimateGas.go (hunk 5 above)
Comment (tudor-malene) [reply go/enclave/rpc/EstimateGas.go:50]: we need to support retrieval by hash. Currently it only works by height

File: go/enclave/rpc/GetTransaction.go (hunk 6 above)
Comment (tudor-malene) [reply go/enclave/rpc/GetTransaction.go:59]: currently we have a static configured base fee. 
Eventually, it needs to be dynamic. 
The problem here was that this was returning 0

File: tools/faucet/faucet/faucet.go [hunk 7]
```diff
@@ -85,7 +83,7 @@ func (f *Faucet) validateTx(tx *types.Transaction) error {
 	for now := time.Now(); time.Since(now) < _timeout; time.Sleep(time.Second) {
 		receipt, err := f.client.TransactionReceipt(context.Background(), tx.Hash())
 		if err != nil {
-			if errors.Is(err, rpc.ErrNilResponse) {
+			if receipt == nil {
```
Comment (BedrockSquirrel) [inline tools/faucet/faucet/faucet.go:86]: You still need to check the type of error here is `ethereum.NotFound` I think? Else it'll keep waiting when the node is down for example.

**Review by BedrockSquirrel — COMMENTED** [review]
Looks reasonable to me so long as tests are all behaving. One minor thing I think needs fixing.

File: tools/faucet/faucet/faucet.go (hunk 7 above)
Comment (tudor-malene) [reply tools/faucet/faucet/faucet.go:86]: good spot. this logic is dodgy here

**Review by BedrockSquirrel — APPROVED** [review]

**Pull request merged**